File: utils/levels.go
package utils

// XP requirement grows by 20% compounded per prestige
func GetXPForLevel(level int, prestige int) int64 {
	rank, ok := Ranks[level]
	if !ok {
		return 0
	}
	base := int64(rank.XPRequired)
	if prestige <= 0 {
		return base
	}
	// multiply by 1.2^prestige
	val := float64(base)
	for i := 0; i < prestige; i++ {
		val *= 1.2
	}
	return int64(val)
}

// GetUserLevel returns highest level achieved at given xp and prestige
func GetUserLevel(xp int64, prestige int) int {
	level := 0
	for l := 0; l < len(Ranks); l++ {
		req := GetXPForLevel(l, prestige)
		if xp >= req {
			level = l
		} else {
			break
		}
	}
	return level
}

File: python/to-do/levels.py
from .constants import RANKS

def get_xp_for_level(level: int, prestige: int) -> int:
    """Calculates the XP required for a given level, factoring in prestige."""
    base_xp = RANKS[level]["xp_required"]
    if prestige == 0:
        return base_xp
    # Increase XP requirement by 20% for each prestige level, compounded
    return int(base_xp * (1.2 ** prestige))

def get_user_level(xp: int, prestige: int) -> int:
    """Calculates a user's level based on their current XP and prestige."""
    level = 0
    for lvl in sorted(RANKS.keys()):
        required_xp = get_xp_for_level(lvl, prestige)
        if xp >= required_xp:
            level = lvl
        else:
            break
    return level
